src/codegen/events/substrate.rs

```diff
@@ -32,68 +45,121 @@ impl EventEmitter for SubstrateEventEmitter<'_> {
-
-                match ty {
-                    Type::String | Type::DynamicBytes => {
-                        let e = expression(
-                            &ast::Expression::Builtin(
-                                pt::Loc::Codegen,
-                                vec![Type::Bytes(32)],
-                                ast::Builtin::Keccak256,
-                                vec![arg.clone()],
-                            ),
-                            cfg,
-                            contract_no,
-                            Some(func),
-                            self.ns,
-                            vartab,
-                            opt,
-                        );
-
-                        topics.push(e);
-                        topic_tys.push(Type::Bytes(32));
-                    }
-                    Type::Struct(_) | Type::Array(..) => {
-                        // We should have an AbiEncodePackedPad
-                        let e = expression(
-                            &ast::Expression::Builtin(
-                                pt::Loc::Codegen,
-                                vec![Type::Bytes(32)],
-                                ast::Builtin::Keccak256,
-                                vec![ast::Expression::Builtin(
-                                    pt::Loc::Codegen,
-                                    vec![Type::DynamicBytes],
-                                    ast::Builtin::AbiEncodePacked,
-                                    vec![arg.clone()],
-                                )],
-                            ),
-                            cfg,
-                            contract_no,
-                            Some(func),
-                            self.ns,
-                            vartab,
-                            opt,
-                        );
-
-                        topics.push(e);
-                        topic_tys.push(Type::Bytes(32));
-                    }
-                    _ => {
-                        let e = expression(arg, cfg, contract_no, Some(func), self.ns, vartab, opt);
-
-                        topics.push(e);
-                        topic_tys.push(ty);
-                    }
-                }
-            } else {
-                let e = expression(arg, cfg, contract_no, Some(func), self.ns, vartab, opt);
-
-                data.push(e);
-                data_tys.push(arg.ty());
+        let loc = pt::Loc::Builtin;
+        let event = &self.ns.events[self.event_no];
+        let hash_len = || Box::new(Expression::NumberLiteral(loc, Type::Uint(32), 32.into()));
+
+        // Events that are not anonymous always have themselves as a topic.
+        // This is static and can be calculated at compile time.
+        if !event.anonymous {
+            let mut encoded =
+                format!("{}::{}", &self.ns.contracts[contract_no].name, &event.name).encode();
+            encoded[0] = 0; // Set the prefix (there is no prefix for the event topic)
```

Comment:
Is the the same as:
```rust
  let encoded = format!("\0{}::{}", &self.ns.contracts[contract_no].name, &event.name).as_bytes();
``` 
?

Reply:
Might be, but I'd prefer to have it explicit here.

Reply:
Aren't you losing the information saved in the first index of this array? If you wish to add another element, you should have done `encoded.insert(0, 0);`

Reply:
> Aren't you losing the information saved in the first index of this array? If you wish to add another element, you should have done encoded.insert(0, 0);

That's correct, but it was intentional. However I came to the conclusion that @seanyoung suggestion is actually better, as it crafts the topic contents directly instead of abusing an encoded string.

Reply:
I don't think @seanyoung solution does the same thing. He added a new character to the string, didn't he?

Reply:
> I don't think @seanyoung's solution does the same thing. He added a new character to the string, didn't he?

Prior, the first byte was the length, which was not needed and replaced with a null byte. But @seanyoung solution is better because it will not depend on the scale encode implementation of `String` to produce correct results.